Sources/PenguinStructures/Tuple.swift

```diff
@@ -0,0 +1,158 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+/// Generalized algebraic product types
+///
+/// Swift's built-in tuple types are algebraic product types, but since they are
+/// not nominal and not easily decomposed, they don't lend themselves to many
+/// types of useful processing.  Models of `TupleProtocol` don't have those
+/// problems.
+public protocol TupleProtocol {
+  /// The type of the first element.
+  associatedtype Head
+
+  /// An algebriac product formed by composing the types of all but the first
+  /// element.
+  associatedtype Tail: TupleProtocol
+
+  /// The first element.
+  var head: Head { get set }
+  
+  /// All elements but the first.
+  var tail: Tail { get set }
+
+  /// The number of elements
+  static var count: Int { get }
+}
+
+extension Empty: TupleProtocol {
+  /// The first element, when `self` is viewed as an instance of algebraic
+  /// product type.
+  public var head: Never { get {  fatalError() } set {  } }
+  
+  /// All elements but the first, when `self` is viewed as an instance of
+  /// algebraic product type.
+  public var tail: Self { get { self } set { } }
+
+  /// The number of elements, when `self` is viewed as an instance of
+  /// algebraic product type.
+  public static var count: Int { 0 }
+}
+
+/// An algebraic product type whose first element is of type `Head` and
+/// whose remaining elements can be stored in `Tail`.
+public struct Tuple<Head, Tail: TupleProtocol>: TupleProtocol {
+  /// The first element.
+  public var head: Head
+  
+  /// All elements but the first.
+  public var tail: Tail
+  
+  /// The number of elements
+  public static var count: Int { Tail.count + 1 }
+}
+
+extension Tuple: DefaultInitializable
+  where Head: DefaultInitializable, Tail: DefaultInitializable
+{
+  // Initialize `self`.
+  public init() {
+    head = Head()
+    tail = Tail()
+  }
+}
+
+extension Tuple: Equatable where Head: Equatable, Tail: Equatable {}
+extension Tuple: Hashable where Head: Hashable, Tail: Hashable {}
+extension Tuple: Comparable where Head: Comparable, Tail: Comparable {
+  public static func < (lhs: Self, rhs: Self) -> Bool {
+    if lhs.head < rhs.head { return true }
+    if lhs.head > rhs.head { return false }
+    return lhs.tail < rhs.tail
+  }
+}
+
+private let prefixLength = "Tuple(".count
+
+extension Tuple: CustomStringConvertible {
+  public var description: String {
+    if Tail.self == Empty.self {
+      return "Tuple(\(String(reflecting:head )))"
+    }
+    else {
+      return "Tuple(\(String(reflecting: head)), "
+        + String(reflecting: tail).dropFirst(prefixLength)
+    }
+  }
+}
+
+// ======== Conveniences ============
+
+public typealias Tuple0 = Empty
```

Comment:
Might we want to add doc comments to these in some form?

Reply:
Yep, good plan. #40

Reply:
#40